File: macos/Sources/CursorRemain/AppUpdater.swift
import AppKit
import CursorTokenCore
import Foundation

enum AppUpdater {
    private static var busy = false

    static var isBusy: Bool { busy }

    static var canSelfUpdate: Bool {
        AppUpdate.looksPackagedMacos(Bundle.main.bundlePath)
    }

    @MainActor
    static func run(
        store: AppStore,
        manual: Bool,
        confirmApply: ((String) -> Bool)? = nil
    ) async -> String {
        if busy { return "正在检查更新…" }
        let cfg = store.config
        if !manual && !cfg.autoUpdateEnabled { return "" }
        if !manual && !canSelfUpdate { return "" }
        if !AppUpdate.shouldAutoCheck(lastCheckAt: cfg.updateLastCheckAt, manual: manual) {
            return cfg.updateLastError.isEmpty ? "已是最新版本" : cfg.updateLastError
        }
        busy = true
        defer { busy = false }
        do {
            let release = try await fetchLatest()
            let decision = AppUpdate.evaluate(
                release: release,
                assetName: AppUpdate.macosAssetName,
                currentSha: AppUpdate.currentCommitSha,
                installedSha: cfg.updateInstalledSha,
                installedAssetId: cfg.updateInstalledAssetId,
                currentVersion: AppUpdate.productVersion
            )
            rememberCheck(store: store, error: decision.upToDate ? "" : (decision.available ? "" : decision.message))
            if decision.upToDate { return decision.message }
            guard decision.available, let asset = decision.asset else {
                return decision.message.isEmpty ? "检查更新失败" : decision.message
            }
            if !canSelfUpdate {
                if manual { openDownloadPage(decision.release?.pageUrl) }
                return decision.message + "。请从下载页安装"
            }
            if manual, let confirmApply, !confirmApply(decision.message) {
                return "已取消更新"
            }
            let staged = try await downloadAndStage(asset)
            if !(await launchHelper(newApp: staged, expectedSha: release.commitSha)) {
                let fail = "已下载更新，但无法启动安装脚本"
                rememberCheck(store: store, error: fail)
                return fail
            }
            if let pending = AppUpdate.rememberedInstallAfterHelper(
                sha: release.commitSha,
                assetId: asset.id,
                helperStarted: true
            ) {
                AppUpdate.writePendingInstall(directory: pendingDirectory(store), sha: pending.sha, assetId: pending.assetId)
            }
            NSApp.terminate(nil)
            return decision.message + "，即将重启"
        } catch is CancellationError {
            return "已取消更新"
        } catch {
            let msg = "检查更新失败: \((error as? CursorAPIError)?.message ?? error.localizedDescription)"
            rememberCheck(store: store, error: msg)
            if manual && msg.contains("无法检查更新") {
                openDownloadPage()
            }
            return msg
        }
    }

    static func openDownloadPage(_ raw: String? = nil) {
        let url = URL(string: raw ?? "") ?? AppUpdate.officialLatestPageURL
        NSWorkspace.shared.open(url)
    }

    private static func fetchLatest() async throws -> AppRelease {
        var apiError: Error?
        var official: AppRelease?
        var rolling: AppRelease?
        do {
            official = try await fetchApiRelease(AppUpdate.apiOfficialLatestURL)
        } catch {
            if let status = (error as? CursorAPIError)?.statusCode, status != 404, !AppUpdate.shouldFallbackFromApi(status), status > 0 {
                throw error
            }
            if let status = (error as? CursorAPIError)?.statusCode, status != 404 {
                apiError = error
            }
        }
        do {
            var release = try await fetchApiRelease(AppUpdate.apiLatestReleaseURL)
            if release.commitSha.isEmpty {
                if let sha = try? await getText(AppUpdate.apiLatestRefURL, accept: "application/vnd.github+json", timeout: 8) {
                    let parsed = AppUpdate.parseTagRefSha(sha)
                    if !parsed.isEmpty { release.commitSha = parsed }
                }
            }
            rolling = release
        } catch {
            if apiError == nil, let status = (error as? CursorAPIError)?.statusCode, status != 404 {
                apiError = error
            } else if apiError == nil {
                apiError = error
            }
        }
        if let chosen = AppUpdate.chooseRelease(official: official, rolling: rolling, currentVersion: AppUpdate.productVersion) {
            return chosen
        }
        do {
            do {
                let html = try await getText(AppUpdate.officialLatestPageURL, accept: "text/html", timeout: 20)
                return try AppUpdate.parseReleasePage(html)
            } catch {
                let html = try await getText(AppUpdate.latestReleasePageURL, accept: "text/html", timeout: 20)
                return try AppUpdate.parseReleasePage(html)
            }
        } catch {
            throw CursorAPIError("无法检查更新（\(shortError(apiError ?? error))）。也可打开下载页手动安装")
        }
    }

    private static func fetchApiRelease(_ url: URL) async throws -> AppRelease {
        var release = try await getJSON(url, parse: AppUpdate.parseRelease, timeout: 10)
        if release.assets.isEmpty {
            release.assets = AppUpdate.knownAssets(release.tag)
        }
        return release
    }

    private static func downloadAndStage(_ asset: AppReleaseAsset) async throws -> URL {
        guard AppUpdate.isAllowedDownloadURL(asset.url), let url = URL(string: asset.url) else {
            throw CursorAPIError("更新地址无效")
        }
        let root = FileManager.default.temporaryDirectory
            .appendingPathComponent("CursorRemain-update-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        let zip = root.appendingPathComponent(AppUpdate.macosAssetName)
        var req = URLRequest(url: url)
        req.setValue(AppUpdate.userAgent, forHTTPHeaderField: "User-Agent")
        req.setValue("application/octet-stream", forHTTPHeaderField: "Accept")
        let (temp, resp) = try await URLSession.shared.download(for: req)
        if let http = resp as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw CursorAPIError("下载失败 HTTP \(http.statusCode)")
        }
        if let final = resp.url, !AppUpdate.isAllowedDownloadURL(final.absoluteString) {
            throw CursorAPIError("更新地址无效")
        }
        let size = (try? FileManager.default.attributesOfItem(atPath: temp.path)[.size] as? NSNumber)?.int64Value ?? 0
        if size > AppUpdate.maxZipBytes { throw CursorAPIError("安装包过大，已取消更新") }
        if FileManager.default.fileExists(atPath: zip.path) {
            try FileManager.default.removeItem(at: zip)
        }
        try FileManager.default.moveItem(at: temp, to: zip)
        let extract = root.appendingPathComponent("extract", isDirectory: true)
        try FileManager.default.createDirectory(at: extract, withIntermediateDirectories: true)
        try await Task.detached(priority: .userInitiated) {
            try Self.run("/usr/bin/ditto", ["-x", "-k", zip.path, extract.path])
        }.value
        guard let app = findApp(in: extract) else {
            throw CursorAPIError("安装包里没有 CursorRemain.app")
        }
        return app
    }

    private static func findApp(in directory: URL) -> URL? {
        let fm = FileManager.default
        guard let en = fm.enumerator(at: directory, includingPropertiesForKeys: [.isDirectoryKey]) else { return nil }
        for case let url as URL in en {
            if url.pathExtension.lowercased() == "app",
               AppUpdate.isOurMacApp(url.lastPathComponent) {
                return url
            }
        }
        return nil
    }

    @MainActor
    static func confirmPending(store: AppStore) {
        let dir = pendingDirectory(store)
        guard let pending = AppUpdate.readPendingInstall(directory: dir) else { return }
        if let confirmed = AppUpdate.confirmPendingInstall(currentSha: AppUpdate.currentCommitSha, pending: pending) {
            AppUpdate.clearPendingInstall(directory: dir)
            rememberInstalled(store: store, sha: confirmed.sha, assetId: confirmed.assetId)
            return
        }
        if AppUpdate.pendingInstallFailed(currentSha: AppUpdate.currentCommitSha, pending: pending) {
            AppUpdate.clearPendingInstall(directory: dir)
            store.config = ConfigStore.update(from: store.settingsDirectory) { live in
                live.updateLastCheckAt = ""
                live.updateLastError = "上次更新没有替换成功，请再试一次"
            }
        }
    }

    @MainActor
    private static func pendingDirectory(_ store: AppStore) -> URL {
        store.settingsDirectory ?? AppPaths.configDirectory()
    }

    private static func launchHelper(newApp: URL, expectedSha: String) async -> Bool {
        let dest = Bundle.main.bundleURL
        let script = FileManager.default.temporaryDirectory
            .appendingPathComponent("CursorRemain-apply-\(UUID().uuidString).sh")
        let log = AppPaths.logPath().deletingLastPathComponent().appendingPathComponent("CursorRemain-update-helper.log")
        let body = """
        #!/bin/bash
        PID="$1"
        SRC="$2"
        DST="$3"
        EXPECT="$4"
        echo "start $(date -u +%Y-%m-%dT%H:%M:%SZ) pid=$PID dst=$DST expect=$EXPECT"
        for i in $(seq 1 80); do
          if ! kill -0 "$PID" 2>/dev/null; then
            break
          fi
          sleep 0.25
        done
        if kill -0 "$PID" 2>/dev/null; then
          kill -9 "$PID" 2>/dev/null || true
          sleep 0.3
        fi
        sleep 0.4
        apply() {
          TMP="${DST}.updating"
          rm -rf "$TMP"
          /usr/bin/ditto "$SRC" "$TMP" || return 1
          if [ -d "$DST" ]; then
            rm -rf "${DST}.old"
            mv "$DST" "${DST}.old" || rm -rf "$DST" || return 1
          fi
          mv "$TMP" "$DST" || return 1
          xattr -cr "$DST" >/dev/null 2>&1 || true
          rm -rf "${DST}.old"
          if [ -n "$EXPECT" ]; then
            GOT=$(/usr/libexec/PlistBuddy -c 'Print :GitCommit' "$DST/Contents/Info.plist" 2>/dev/null | tr '[:upper:]' '[:lower:]')
            EXP=$(printf '%s' "$EXPECT" | tr '[:upper:]' '[:lower:]')
            case "$GOT" in
              "$EXP"*) ;;
              *)
                case "$EXP" in
                  "$GOT"*) ;;
                  *) echo "sha mismatch got=$GOT expect=$EXP"; return 1 ;;
                esac
                ;;
            esac
          fi
          return 0
        }
        ok=0
        for t in 1 2 3 4 5; do
          if apply; then ok=1; break; fi
          echo "retry $t"
          sleep 0.4
        done
        if [ "$ok" -ne 1 ]; then
          echo "apply failed"
          rm -f "$0"
          exit 1
        fi
        EXE="$DST/Contents/MacOS/CursorRemain"
        if [ ! -x "$EXE" ]; then
          EXE="$DST/Contents/MacOS/CursorTokenTray"
        fi
        if [ -x "$EXE" ]; then
          /usr/bin/nohup "$EXE" >/dev/null 2>&1 &
        else
          /usr/bin/open -n -g "$DST"
        fi
        echo "opened $DST"
        rm -f "$0"
        """
        do {
            AppPaths.ensureDirectory(log.deletingLastPathComponent())
            try body.write(to: script, atomically: true, encoding: .utf8)
            try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: script.path)
            let command = AppUpdate.detachedHelperLaunchCommand(
                script: script.path,
                pid: String(ProcessInfo.processInfo.processIdentifier),
                source: newApp.path,
                destination: dest.path,
                expectedSha: AppUpdate.normalizeSha(expectedSha),
                log: log.path
            )
            let status = try await Task.detached(priority: .userInitiated) { () -> Int32 in
                let proc = Process()
                proc.executableURL = URL(fileURLWithPath: "/bin/bash")
                proc.arguments = ["-c", command]
                proc.standardOutput = FileHandle.nullDevice
                proc.standardError = FileHandle.nullDevice
                try proc.run()
                proc.waitUntilExit()
                return proc.terminationStatus
            }.value
            if status != 0 {
                AppLog.log("无法启动更新脚本: exit \(status)")
                return false
            }
            return true
        } catch {
            AppLog.log("无法启动更新脚本: \(error.localizedDescription)")
            return false
        }
    }

    @MainActor
    private static func rememberCheck(store: AppStore, error: String) {
        store.config = ConfigStore.update(from: store.settingsDirectory) { live in
            live.updateLastCheckAt = AppUpdate.nowIso()
            live.updateLastError = error
        }
    }

    @MainActor
    private static func rememberInstalled(store: AppStore, sha: String, assetId: Int64) {
        store.config = ConfigStore.update(from: store.settingsDirectory) { live in
            live.updateLastCheckAt = AppUpdate.nowIso()
            live.updateLastError = ""
            live.updateInstalledSha = AppUpdate.normalizeSha(sha)
            live.updateInstalledAssetId = assetId
        }
    }

    private static func getJSON(_ url: URL, parse: (String) throws -> AppRelease, timeout: TimeInterval) async throws -> AppRelease {
        try parse(try await getText(url, accept: "application/vnd.github+json", timeout: timeout))
    }

    private static func getText(_ url: URL, accept: String, timeout: TimeInterval) async throws -> String {
        var req = URLRequest(url: url, timeoutInterval: timeout)
        req.setValue(AppUpdate.userAgent, forHTTPHeaderField: "User-Agent")
        req.setValue(accept, forHTTPHeaderField: "Accept")
        if accept.contains("github") {
            req.setValue("2022-11-28", forHTTPHeaderField: "X-GitHub-Api-Version")
        }
        let (data, resp) = try await URLSession.shared.data(for: req)
        if let http = resp as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw CursorAPIError(AppUpdate.httpStatusMessage(http.statusCode), statusCode: http.statusCode)
        }
        return String(data: data, encoding: .utf8) ?? ""
    }

    private static func shortError(_ error: Error) -> String {
        if let api = error as? CursorAPIError { return api.message }
        let ns = error as NSError
        if ns.domain == NSURLErrorDomain && ns.code == NSURLErrorTimedOut { return "连接超时" }
        return error.localizedDescription
    }

    private static func run(_ launchPath: String, _ arguments: [String]) throws {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: launchPath)
        proc.arguments = arguments
        proc.standardOutput = FileHandle.nullDevice
        proc.standardError = FileHandle.nullDevice
        try proc.run()
        proc.waitUntilExit()
        if proc.terminationStatus != 0 {
            throw CursorAPIError("解压更新失败")
        }
    }
}

File: windows/CursorRemain/AppUpdater.cs
using System.Diagnostics;
using System.Net.Http.Headers;
using CursorTokenCore;

namespace CursorRemain;

static class AppUpdater
{
    static readonly HttpClient Http = CreateClient();
    static bool _busy;

    static HttpClient CreateClient()
    {
        var http = new HttpClient(new HttpClientHandler { AllowAutoRedirect = true, UseCookies = false })
        {
            Timeout = TimeSpan.FromMinutes(3),
        };
        http.DefaultRequestHeaders.UserAgent.ParseAdd(AppUpdate.UserAgent);
        http.DefaultRequestHeaders.Accept.ParseAdd("application/vnd.github+json");
        return http;
    }

    public static bool Busy => _busy;

    public static bool CanSelfUpdate =>
        AppUpdate.LooksPackagedWindows(Environment.ProcessPath ?? Application.ExecutablePath);

    public static async Task<string> RunAsync(
        AppConfig cfg,
        bool manual,
        Func<string, bool>? confirmApply,
        Action? restart,
        CancellationToken ct = default)
    {
        if (_busy) return "正在检查更新…";
        if (!manual && !cfg.AutoUpdateEnabled) return "";
        if (!manual && !CanSelfUpdate) return "";
        if (!AppUpdate.ShouldAutoCheck(cfg.UpdateLastCheckAt, DateTimeOffset.UtcNow, manual))
            return string.IsNullOrWhiteSpace(cfg.UpdateLastError) ? "已是最新版本" : cfg.UpdateLastError;

        _busy = true;
        try
        {
            var release = await FetchLatest(ct);
            var preferLight = AppUpdate.CanUseWindowsLightUpdate();
            var decision = AppUpdate.Evaluate(
                release,
                AppUpdate.PreferredWindowsAssetName(preferLight),
                AppUpdate.CurrentCommitSha(),
                cfg.UpdateInstalledSha,
                cfg.UpdateInstalledAssetId,
                AppUpdate.ProductVersion);
            RememberCheck(cfg, decision.UpToDate ? "" : (decision.Available ? "" : decision.Message));
            if (decision.UpToDate) return decision.Message;
            if (!decision.Available || decision.Asset is null)
                return string.IsNullOrWhiteSpace(decision.Message) ? "检查更新失败" : decision.Message;

            if (!CanSelfUpdate)
            {
                if (manual) OpenDownloadPage(decision.Release?.PageUrl);
                return decision.Message + "。请从下载页安装";
            }

            if (manual && confirmApply is not null && !confirmApply(decision.Message))
                return "已取消更新";

            var asset = decision.Asset;
            string applied;
            try
            {
                applied = await DownloadAndStage(asset, ct);
            }
            catch (Exception ex) when (
                preferLight
                && AppUpdate.IsWindowsLightAssetName(asset.Name)
                && ex is not OperationCanceledException)
            {
                CrashLog.Write(ex);
                var fallback = AppUpdate.FindPreferredAsset(release, AppUpdate.WindowsAssetName);
                if (fallback is null || string.Equals(fallback.Url, asset.Url, StringComparison.OrdinalIgnoreCase))
                    throw;
                applied = await DownloadAndStage(fallback, ct);
                asset = fallback;
            }
            if (!LaunchHelper(applied))
            {
                var fail = "已下载更新，但无法启动安装脚本";
                RememberCheck(cfg, fail);
                return fail;
            }
            if (AppUpdate.RememberedInstallAfterHelper(release.CommitSha, asset.Id, true) is { } pending)
                AppUpdate.WritePendingInstall(AppPaths.ConfigDirectory(), pending.Sha, pending.AssetId);
            restart?.Invoke();
            return decision.Message + "，即将重启";
        }
        catch (OperationCanceledException)
        {
            return "已取消更新";
        }
        catch (Exception ex)
        {
            var msg = "检查更新失败: " + ex.Message;
            RememberCheck(cfg, msg);
            if (manual && msg.Contains("无法检查更新", StringComparison.Ordinal))
                OpenDownloadPage();
            return msg;
        }
        finally
        {
            _busy = false;
        }
    }

    public static void OpenDownloadPage(string? url = null)
    {
        var target = string.IsNullOrWhiteSpace(url) ? AppUpdate.OfficialLatestPageUrl : url;
        try { Process.Start(new ProcessStartInfo(target) { UseShellExecute = true }); }
        catch (Exception ex) { CrashLog.Write(ex); }
    }

    static async Task<AppRelease> FetchLatest(CancellationToken ct)
    {
        Exception? apiError = null;
        AppRelease? official = null;
        AppRelease? rolling = null;
        try
        {
            official = await TryFetchApi(AppUpdate.ApiOfficialLatestUrl, ct, TimeSpan.FromSeconds(10));
        }
        catch (Exception ex) when (ex is not OperationCanceledException || !ct.IsCancellationRequested)
        {
            apiError = ex;
        }
        try
        {
            rolling = await TryFetchApi(AppUpdate.ApiLatestReleaseUrl, ct, TimeSpan.FromSeconds(10));
            if (rolling is { CommitSha.Length: 0 })
            {
                try
                {
                    using var apiCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
                    apiCts.CancelAfter(TimeSpan.FromSeconds(8));
                    using var refReq = new HttpRequestMessage(HttpMethod.Get, AppUpdate.ApiLatestRefUrl);
                    refReq.Headers.TryAddWithoutValidation("X-GitHub-Api-Version", "2022-11-28");
                    using var refResp = await Http.SendAsync(refReq, apiCts.Token);
                    if (refResp.IsSuccessStatusCode)
                    {
                        var sha = AppUpdate.ParseTagRefSha(await refResp.Content.ReadAsStringAsync(apiCts.Token));
                        if (sha.Length > 0)
                            rolling = WithSha(rolling, sha);
                    }
                }
                catch (Exception ex) { CrashLog.Write(ex); }
            }
        }
        catch (Exception ex) when (ex is not OperationCanceledException || !ct.IsCancellationRequested)
        {
            apiError ??= ex;
        }

        var chosen = AppUpdate.ChooseRelease(official, rolling, AppUpdate.ProductVersion);
        if (chosen is not null) return chosen;

        try
        {
            using var pageCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            pageCts.CancelAfter(TimeSpan.FromSeconds(20));
            using var req = new HttpRequestMessage(HttpMethod.Get, AppUpdate.OfficialLatestPageUrl);
            req.Headers.Accept.Clear();
            req.Headers.Accept.ParseAdd("text/html");
            using var resp = await Http.SendAsync(req, pageCts.Token);
            var html = await resp.Content.ReadAsStringAsync(pageCts.Token);
            if (resp.IsSuccessStatusCode)
                return AppUpdate.ParseReleasePage(html);
            using var latestReq = new HttpRequestMessage(HttpMethod.Get, AppUpdate.LatestReleasePageUrl);
            latestReq.Headers.Accept.Clear();
            latestReq.Headers.Accept.ParseAdd("text/html");
            using var latestResp = await Http.SendAsync(latestReq, pageCts.Token);
            html = await latestResp.Content.ReadAsStringAsync(pageCts.Token);
            if (!latestResp.IsSuccessStatusCode)
                throw new InvalidOperationException(AppUpdate.HttpStatusMessage((int)latestResp.StatusCode));
            return AppUpdate.ParseReleasePage(html);
        }
        catch (Exception pageError) when (apiError is not null)
        {
            throw new InvalidOperationException(
                "无法检查更新（" + ShortError(apiError) + "）。也可打开下载页手动安装",
                pageError);
        }
        throw new InvalidOperationException("无法检查更新。也可打开下载页手动安装");
    }

    static async Task<AppRelease?> TryFetchApi(string url, CancellationToken ct, TimeSpan timeout)
    {
        using var apiCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        apiCts.CancelAfter(timeout);
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.TryAddWithoutValidation("X-GitHub-Api-Version", "2022-11-28");
        using var resp = await Http.SendAsync(req, apiCts.Token);
        var body = await resp.Content.ReadAsStringAsync(apiCts.Token);
        if (resp.IsSuccessStatusCode)
            return FillAssets(AppUpdate.ParseRelease(body));
        if ((int)resp.StatusCode == 404) return null;
        if (!AppUpdate.ShouldFallbackFromApi((int)resp.StatusCode))
            throw new InvalidOperationException(AppUpdate.HttpStatusMessage((int)resp.StatusCode));
        throw new InvalidOperationException(AppUpdate.HttpStatusMessage((int)resp.StatusCode));
    }

    static AppRelease FillAssets(AppRelease release)
    {
        if (release.Assets.Count > 0) return release;
        return new AppRelease
        {
            Tag = release.Tag,
            Version = release.Version,
            Body = release.Body,
            CommitSha = release.CommitSha,
            PageUrl = release.PageUrl,
            PublishedAt = release.PublishedAt,
            Assets = AppUpdate.KnownAssets(release.Tag),
        };
    }

    static AppRelease WithSha(AppRelease release, string sha) =>
        new()
        {
            Tag = release.Tag,
            Version = release.Version,
            Body = release.Body,
            CommitSha = sha,
            PageUrl = release.PageUrl,
            PublishedAt = release.PublishedAt,
            Assets = release.Assets,
        };

    static string ShortError(Exception ex) =>
        ex is OperationCanceledException or TaskCanceledException ? "连接超时" : ex.Message;

    static async Task<string> DownloadAndStage(AppReleaseAsset asset, CancellationToken ct)
    {
        if (!AppUpdate.IsAllowedDownloadUrl(asset.Url))
            throw new InvalidOperationException("更新地址无效");
        var root = Path.Combine(Path.GetTempPath(), "CursorRemain-update-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(root);
        var zipName = Path.GetFileName(asset.Name);
        if (string.IsNullOrWhiteSpace(zipName) || zipName.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0)
            zipName = AppUpdate.WindowsAssetName;
        var zip = Path.Combine(root, zipName);
        using (var req = new HttpRequestMessage(HttpMethod.Get, asset.Url))
        {
            req.Headers.Accept.Clear();
            req.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/octet-stream"));
            using var resp = await Http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
            resp.EnsureSuccessStatusCode();
            if (resp.RequestMessage?.RequestUri is { } final && !AppUpdate.IsAllowedDownloadUrl(final.ToString()))
                throw new InvalidOperationException("更新地址无效");
            var length = resp.Content.Headers.ContentLength;
            if (length > AppUpdate.MaxZipBytes)
                throw new InvalidOperationException("安装包过大，已取消更新");
            await using var input = await resp.Content.ReadAsStreamAsync(ct);
            await using var output = File.Create(zip);
            var buffer = new byte[81920];
            long total = 0;
            while (true)
            {
                var n = await input.ReadAsync(buffer, ct);
                if (n == 0) break;
                total += n;
                if (total > AppUpdate.MaxZipBytes)
                    throw new InvalidOperationException("安装包过大，已取消更新");
                await output.WriteAsync(buffer.AsMemory(0, n), ct);
            }
        }
        var extract = Path.Combine(root, "extract");
        AppUpdate.ExtractZipSafe(zip, extract);
        return AppUpdate.FindExtractedWindowsExe(extract)
            ?? throw new InvalidOperationException("安装包里没有 CursorRemain.exe");
    }

    public static void ConfirmPending(AppConfig cfg)
    {
        var dir = AppPaths.ConfigDirectory();
        var pending = AppUpdate.ReadPendingInstall(dir);
        if (pending is null) return;
        if (AppUpdate.ConfirmPendingInstall(AppUpdate.CurrentCommitSha(), pending) is { } confirmed)
        {
            AppUpdate.ClearPendingInstall(dir);
            RememberInstalled(cfg, confirmed.Sha, confirmed.AssetId);
            return;
        }
        if (AppUpdate.PendingInstallFailed(AppUpdate.CurrentCommitSha(), pending))
        {
            AppUpdate.ClearPendingInstall(dir);
            try
            {
                ConfigStore.Update(live =>
                {
                    live.UpdateLastCheckAt = "";
                    live.UpdateLastError = "上次更新没有替换成功，请再试一次";
                    cfg.UpdateLastCheckAt = "";
                    cfg.UpdateLastError = live.UpdateLastError;
                    cfg.AutoUpdateEnabled = live.AutoUpdateEnabled;
                    cfg.UpdateInstalledSha = live.UpdateInstalledSha;
                    cfg.UpdateInstalledAssetId = live.UpdateInstalledAssetId;
                });
            }
            catch (Exception ex) { CrashLog.Write(ex); }
        }
    }

    static bool LaunchHelper(string newExe)
    {
        var dest = Environment.ProcessPath ?? Application.ExecutablePath;
        if (string.IsNullOrWhiteSpace(dest)) return false;
        var bat = Path.Combine(Path.GetTempPath(), "CursorRemain-apply-" + Guid.NewGuid().ToString("N") + ".cmd");
        var script =
            "@echo off\r\n" +
            "setlocal EnableExtensions\r\n" +
            "set \"PID=%~1\"\r\n" +
            "set \"SRC=%~2\"\r\n" +
            "set \"DST=%~3\"\r\n" +
            "set /a _i=0\r\n" +
            ":wait\r\n" +
            "if %_i% GEQ 80 goto kill\r\n" +
            "ping -n 2 127.0.0.1 >nul\r\n" +
            "tasklist /FI \"PID eq %PID%\" 2>nul | findstr /I /C:\" %PID% \" >nul\r\n" +
            "if not errorlevel 1 (\r\n" +
            "  set /a _i+=1\r\n" +
            "  goto wait\r\n" +
            ")\r\n" +
            "goto copy\r\n" +
            ":kill\r\n" +
            "taskkill /F /PID %PID% >nul 2>&1\r\n" +
            "ping -n 2 127.0.0.1 >nul\r\n" +
            ":copy\r\n" +
            "set /a _c=0\r\n" +
            ":retry\r\n" +
            "copy /Y \"%SRC%\" \"%DST%\" >nul\r\n" +
            "if not errorlevel 1 goto start\r\n" +
            "set /a _c+=1\r\n" +
            "if %_c% GEQ 8 goto fail\r\n" +
            "ping -n 2 127.0.0.1 >nul\r\n" +
            "goto retry\r\n" +
            ":fail\r\n" +
            "exit /b 1\r\n" +
            ":start\r\n" +
            "start \"\" \"%DST%\"\r\n" +
            "del \"%~f0\"\r\n";
        File.WriteAllText(bat, script);
        var psi = new ProcessStartInfo
        {
            FileName = bat,
            Arguments = $"{Environment.ProcessId} \"{newExe}\" \"{dest}\"",
            UseShellExecute = true,
            WindowStyle = ProcessWindowStyle.Hidden,
            WorkingDirectory = Path.GetTempPath(),
        };
        return Process.Start(psi) is not null;
    }

    static void RememberCheck(AppConfig cfg, string error)
    {
        try
        {
            ConfigStore.Update(live =>
            {
                live.UpdateLastCheckAt = AppUpdate.NowIso();
                live.UpdateLastError = error;
                cfg.UpdateLastCheckAt = live.UpdateLastCheckAt;
                cfg.UpdateLastError = live.UpdateLastError;
                cfg.AutoUpdateEnabled = live.AutoUpdateEnabled;
                cfg.UpdateInstalledSha = live.UpdateInstalledSha;
                cfg.UpdateInstalledAssetId = live.UpdateInstalledAssetId;
            });
        }
        catch (Exception ex) { CrashLog.Write(ex); }
    }

    static void RememberInstalled(AppConfig cfg, string sha, long assetId)
    {
        try
        {
            ConfigStore.Update(live =>
            {
                live.UpdateLastCheckAt = AppUpdate.NowIso();
                live.UpdateLastError = "";
                live.UpdateInstalledSha = AppUpdate.NormalizeSha(sha);
                live.UpdateInstalledAssetId = assetId;
                cfg.UpdateLastCheckAt = live.UpdateLastCheckAt;
                cfg.UpdateLastError = "";
                cfg.UpdateInstalledSha = live.UpdateInstalledSha;
                cfg.UpdateInstalledAssetId = live.UpdateInstalledAssetId;
            });
        }
        catch (Exception ex) { CrashLog.Write(ex); }
    }
}
